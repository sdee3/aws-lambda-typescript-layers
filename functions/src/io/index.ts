export * from './io'
export * from './s3-io'
