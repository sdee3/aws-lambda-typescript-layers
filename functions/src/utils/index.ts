export * from './delay'
export * from './globalOverrides'
export * from './s3'
export * from './three'
