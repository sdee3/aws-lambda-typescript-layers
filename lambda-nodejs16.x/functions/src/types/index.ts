import {
  APIGatewayProxyEvent,
  APIGatewayProxyResult,
  Handler,
} from 'aws-lambda'

export type APIGatewayProxyHandler = Handler<
  APIGatewayProxyEvent,
  APIGatewayProxyResult
>

export interface GetS3ObjectByKeyEvent extends APIGatewayProxyEvent {
  key?: string
}
