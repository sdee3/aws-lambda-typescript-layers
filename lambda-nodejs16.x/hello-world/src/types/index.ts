import { APIGatewayProxyEvent } from "aws-lambda";

export interface GetS3ObjectByKeyEvent extends APIGatewayProxyEvent {
  bucket: string;
  key: string;
}
