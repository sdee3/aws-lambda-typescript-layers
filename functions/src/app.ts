import { GetS3ObjectByKeyEvent } from './types'
import {
  convertFbxToGlb,
  getS3ObjectUrl,
  initS3Client,
  overrideGlobal,
  uploadObjectToS3,
} from './utils'

overrideGlobal()

const client = initS3Client()

export const lambdaHandler = async (event: GetS3ObjectByKeyEvent) => {
  if (!event?.key || !event.folder || !process.env.BUCKET_NAME)
    return { statusCode: 400, body: 'Wrong parameters provided!' }

  const Bucket = process.env.BUCKET_NAME
  const Folder = event.folder
  const Key = event.key

  const s3ObjectUrl = await getS3ObjectUrl(client, Bucket, Folder, Key)
  const handlerResponse = await convertFbxToGlb(s3ObjectUrl)

  const outFileName = Key.replace('.fbx', '.glb')

  await uploadObjectToS3(client, Bucket, Folder, outFileName, handlerResponse)

  return {
    statusCode: 200,
    message: 'Successfully uploaded GLB file to S3: ' + outFileName,
  }
}
