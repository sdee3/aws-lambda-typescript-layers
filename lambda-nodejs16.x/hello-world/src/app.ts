import { Stream } from 'stream'
import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3'
import { BoxGeometry, Scene, MeshBasicMaterial, Mesh } from 'three'
import { APIGatewayProxyResult } from 'aws-lambda'

import { APIGatewayProxyHandler, GetS3ObjectByKeyEvent } from './types'

const client = new S3Client({})

const getS3Object = (Bucket: string, Key: string): Promise<string> => {
  return new Promise(async (resolve, reject) => {
    const getObjectCommand = new GetObjectCommand({ Bucket, Key })

    try {
      const response = await client.send(getObjectCommand)

      if (!response?.Body) throw new Error('No response')

      // Store all of data chunks returned from the response data stream
      // into an array then use Array#join() to use the returned contents as a String
      const responseDataChunks: Buffer[] = []

      const responseBodyStream = response.Body as Stream

      // Handle an error while streaming the response body
      responseBodyStream.once('error', (err: Error) => reject(err))

      // Attach a 'data' listener to add the chunks of data to our array
      // Each chunk is a Buffer instance
      responseBodyStream.on('data', (chunk: Buffer) =>
        responseDataChunks.push(chunk)
      )

      // Once the stream has no more data, join the chunks into a string and return the string
      responseBodyStream.once('end', () => resolve(responseDataChunks.join('')))
    } catch (err) {
      // Handle the error or throw
      return reject(err)
    }
  })
}

export const lambdaHandler: APIGatewayProxyHandler = async (
  event: GetS3ObjectByKeyEvent,
  _context
) => {
  if (!event?.bucket || !event?.key)
    return { statusCode: 400, body: 'No bucket or key' }

  let handlerResponse: APIGatewayProxyResult | null = null

  try {
    const data = await getS3Object(event.bucket, event.key)

    const scene = new Scene()
    const geometry = new BoxGeometry(1, 1, 1)
    const material = new MeshBasicMaterial({ color: 0x00ff00 })
    const cube = new Mesh(geometry, material)

    console.info('Scene ID:', scene.id)
    console.info('Geometry ID:', geometry.id)
    console.info('Material ID:', material.id)
    console.info('Cube ID:', cube.id)

    handlerResponse = {
      statusCode: 200,
      body: JSON.stringify({
        message: event.key,
        dataSize: data.length,
      }),
    }
  } catch (err) {
    handlerResponse = {
      statusCode: 500,
      body: JSON.stringify({
        message: err,
      }),
    }
  }

  return handlerResponse
}
