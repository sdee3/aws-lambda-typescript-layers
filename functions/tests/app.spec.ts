/* eslint-disable @typescript-eslint/no-explicit-any */
import { Readable } from 'stream'
import { APIGatewayProxyEvent } from 'aws-lambda'
import { lambdaHandler } from '../src/app'

const objectKey = 'my-object.mbtiles'
const mockStream = Readable.from(objectKey)

jest.mock('@aws-sdk/client-s3', () => {
  const mockS3Client = {
    send: jest
      .fn()
      .mockImplementation(() => Promise.resolve({ Body: mockStream })),
  }

  return { S3Client: jest.fn(() => mockS3Client), GetObjectCommand: jest.fn() }
})

describe('Unit test for app handler', function () {
  it('verifies error code 400 response', async () => {
    const event: APIGatewayProxyEvent = {
      key: '1',
    } as any
    const result = await lambdaHandler(event)

    expect(result.statusCode).toEqual(400)
    expect(result.body).toMatchSnapshot()
  })

  it('verifies successful response', async () => {
    process.env.BUCKET_NAME = 'test-hwp-7331'

    const event = {
      key: objectKey,
    } as any

    const result = await lambdaHandler(event)

    expect(result.statusCode).toEqual(200)

    const parsedBody = JSON.parse(result.body)

    expect(parsedBody).toHaveProperty('message')
    expect(parsedBody).toHaveProperty('dataSize')

    expect(parsedBody.message).toEqual(objectKey)
    expect(parsedBody.dataSize).toEqual(objectKey.length)
  })
})
