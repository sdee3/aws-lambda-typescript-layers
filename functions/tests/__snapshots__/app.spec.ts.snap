// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for app handler catches error 500 with empty body 1`] = `"{\\"message\\":{}}"`;

exports[`Unit test for app handler verifies error code 400 response 1`] = `"No key provided!"`;
